static/js/timerender.ts

```diff
@@ -24,6 +24,91 @@ export function clear_for_testing(): void {
     next_timerender_id = 0;
 }
 
+type DateFormat = "short" | "medium" | "long" | "day" | "full";
+type DateWithTimeFormat = "short12" | "short24" | "medium12" | "medium24" | "long12" | "long24";
+type TimeFormat = "hour12" | "hour24";
+
+type DateOrTimeFormat = DateFormat | TimeFormat | DateWithTimeFormat;
+
+// The function takes in an optional type parameter and returns
+// an options object that can be used to format a date/time using the Intl.DateTimeFormat API.
+// This function allows easy and consistent formatting of dates and times
+// with different options to customize the format according to the specified type
+function get_format_options_for_type(type?: DateOrTimeFormat): Intl.DateTimeFormatOptions {
+    const hour12FormatOptions: Intl.DateTimeFormatOptions = {
+        hourCycle: "h12",
+        hour: "numeric",
+        minute: "2-digit",
+    };
+    const hour24FormatOptions: Intl.DateTimeFormatOptions = {
+        hourCycle: "h23",
+        hour: "2-digit",
+        minute: "2-digit",
+    };
+
+    const dayFormatOptions: Intl.DateTimeFormatOptions = {weekday: "long"};
+    const fullFormatOptions: Intl.DateTimeFormatOptions = {dateStyle: "full"};
+
+    const shortFormatOptions: Intl.DateTimeFormatOptions = {day: "2-digit", month: "short"};
+    const mediumFormatOptions: Intl.DateTimeFormatOptions = {
+        ...shortFormatOptions,
+        year: "numeric",
+    };
+    const longFormatOptions: Intl.DateTimeFormatOptions = {
+        ...mediumFormatOptions,
+        ...hour24FormatOptions,
+        weekday: "short",
+    };
+
+    switch (type) {
+        case "day": // Wednesday
+            return dayFormatOptions;
+        case "full": // Wednesday, July 27, 2016
+            return fullFormatOptions;
+        case "hour12": // 01:30 PM
+            return hour12FormatOptions;
+        case "hour24": // 13:30
+            return hour24FormatOptions;
+        case "short": // Jul 27
+            return shortFormatOptions;
+        case "short12": // Jul 27, 01:30 PM
+            return {...shortFormatOptions, ...hour12FormatOptions};
+        case "short24": // Jul 27, 13:30
+            return {...shortFormatOptions, ...hour24FormatOptions};
+        case "medium": // Jul 27, 2016
+            return mediumFormatOptions;
+        case "medium12": // Jul 27, 2016, 01:30 PM
+            return {...mediumFormatOptions, ...hour12FormatOptions};
+        case "medium24": // Jul 27, 2016, 13:30
+            return {...mediumFormatOptions, ...hour24FormatOptions};
+        case "long": // Wed, Jul 27, 2016, 01:30 PM
+        case "long12": // Wed, Jul 27, 2016, 01:30 PM
+            return {...longFormatOptions, ...hour12FormatOptions};
+        case "long24": // Wed, Jul 27, 2016, 13:30
+            return {...longFormatOptions, ...hour24FormatOptions};
+        default: // Jul 27, 2016
+            return mediumFormatOptions;
+    }
```

Comment:
@alya this block and the tests above are the definitions of the set of time formats we are using in the app. It's worth your reviewing for any of them seem off.

The main thing I've noticed is that we're using 2-digit style for dates, like "Jul 03", rather than "Jul 3". That is potentially desirable in some contexts to provide a more fixed width to dates, but probably in most actual settings it'd be better to drop the leading 0s there too.

We could also think of more descriptive names than short/medium/long :).

Reply:
Yeah, this was discussed on CZO -- dropping the leading 0's makes sense to me.

Reply:
As discussed earlier, I've removed the leading zeros for dates and 12-hour time format. 

I can't think of any better names for the format types than "short", "medium", and "long" at the moment. "Date", "DateYear", "WeekdayDateYear" would be more semantic, but I don't particularly like them. Let me know if you have any ideas.